packages/smooth_app/lib/pages/product/common/product_query_page.dart

```diff
@@ -195,36 +204,62 @@ class _ProductQueryPageState extends State<ProductQueryPage>
           onRefresh: () => refreshList(),
           child: ListView.builder(
             controller: _scrollController,
+            // To allow refresh even when not the whole page is filled
+            physics: const AlwaysScrollableScrollPhysics(),
             itemBuilder: (BuildContext context, int index) {
               if (index == 0) {
                 // on top, a message
                 return _getTopMessagesCard();
               }
               index--;
+
+              final int barcodesCount = _model.displayBarcodes.length;
+
               // TODO(monsieurtanuki): maybe call it earlier, like for first unknown page index - 5?
-              if (index >= _model.displayBarcodes.length) {
+              if (index >= barcodesCount) {
                 _downloadNextPage();
               }
-              if (index >= _model.displayBarcodes.length) {
+
+              if (index >= barcodesCount) {
                 // TODO(monsieurtanuki): maybe display something specific for data being downloaded (the next page) and unknown data (beyond next page)
-                return const SmoothProductCardTemplate();
+
+                // (+1) index starts with 0
+                final int overIndexCount = index + 1 - barcodesCount;
+
+                if (overIndexCount <= OVERSCROLL_TEMPLATE_COUNT) {
+                  return const SmoothProductCardTemplate();
+                } else if (overIndexCount == 1 + OVERSCROLL_TEMPLATE_COUNT) {
+                  return const Center(child: CircularProgressIndicator());
+                } else if (overIndexCount > 1 + OVERSCROLL_TEMPLATE_COUNT) {
+                  return SizedBox(
+                    height: MediaQuery.of(context).size.height / 4,
+                  );
+                }
```

Comment:
I can't really understand this piece. When are the three cases used? In particular, when is the template card displayed and when is the circular progress displayed?